File: 1213-intersection-of-three-sorted-arrays/1213.py
class Solution:
    def bsearch(self, arr, target):
        head = 0
        tail = len(arr) - 1

        while head <= tail:
            mid = head + (tail - head) // 2
            if arr[mid] == target:
                return True
            elif arr[mid] < target:
                head = mid + 1
            else:
                tail = mid - 1

        return False

    def arraysIntersection(self, arr1: List[int], arr2: List[int], arr3: List[int]) -> List[int]:
        ans = []
        for a in arr1:
           if self.bsearch(arr2, a) and self.bsearch(arr3, a):
               ans.append(a)

        return ans

File: 1213-intersection-of-three-sorted-arrays/1213.cpp
class Solution {
private:
    bool exist(vector<int>& arr, int val) {
        int left = -1;
        int right = arr.size() - 1;
        while (left < right) {
            int mid = (left + right + 1) / 2;
            if (arr[mid] == val)
                return true;
            else if (arr[mid] < val)
                left = mid;
            else
                right = mid - 1;
        }
        return false;
    }
public:
    vector<int> arraysIntersection(vector<int>& arr1, vector<int>& arr2, vector<int>& arr3) {
        vector<int> ans;
        for (int val : arr1) {
            if (exist(arr2, val) && exist(arr3, val))
                ans.push_back(val);
        }
        return ans;
    }
};
